packages/circuit-ui/components/Dialog/Dialog.tsx

```diff
@@ -258,17 +268,26 @@ export const Dialog = forwardRef<HTMLDialogElement, DialogProps>(
     }, [animationDurationRef.current, onCloseStart]);
 
     const handleOutsideClick = useCallback(() => {
-      lastFocusedElementRef.current = null;
-      handleDialogClose();
-    }, [handleDialogClose]);
+      if (!preventOutsideClickClose) {
+        lastFocusedElementRef.current = null;
+        handleDialogClose();
+      }
+    }, [handleDialogClose, preventOutsideClickClose]);
 
     const useClickOutsideRefs = preventOutsideClickRefs
       ? // eslint-disable-next-line compat/compat
         [dialogRef, preventOutsideClickRefs].flat()
       : [dialogRef];
 
-    useClickOutside(useClickOutsideRefs, handleOutsideClick, open && !isModal);
-    useEscapeKey(() => handleDialogClose(), open && !isModal);
+    useClickOutside(
+      useClickOutsideRefs,
+      handleOutsideClick,
+      open && !preventOutsideClickClose,
+    );
+    useEscapeKey((e: KeyboardEvent) => {
+      e.preventDefault();
+      handleDialogClose();
+    }, open && !preventEscapeKeyClose);
```

Comment:
This inline arrow function changes on every render, which means the event listener inside `useEscapeKey` will be removed and re-added on every render. Wrap it in `useCallback` to avoid that.